src/cljs/s3_beam/client.cljs

```diff
@@ -31,28 +65,80 @@
-        signature  (select-keys (:signature upload-info) sig-fields)
-        form-data  (formdata-from-map (merge signature {:file (:f upload-info)}))]
-    (xhr/send
-     (:action (:signature upload-info))
-     (fn [res]
-       (let [loc (aget (.getElementsByTagName (.getResponseXml (.-target res)) "Location") 0)]
-         (put! ch (gdom/getTextContent loc))
-         (close! ch)))
-     "POST"
-     form-data)))
+(defn xml->map
+  "Transforms a shallow XML document into a map."
+  [xml]
+  (let [kv (->> (array-seq (.. xml -firstChild -childNodes))
+                (map (fn [n] [(.toLowerCase (.-nodeName n))
+                             (gdom/getTextContent n)])))]
+    (zipmap (map (comp keyword first) kv) (map second kv))))
+
+(defn upload-file
+  "Take the signature information `upload-info`, and a channel `ch`,
+  and does the S3 upload.
+  Returns a map with:
+    :file - the File that was uploaded
+    :response - a map of S3 metadata: :key, :location, :bucket, :etag
+    :identifier - an identifier pass-through value for caller to identify the file with
+  If the upload-info has an :error-code key because the signing failed, skips the xhr/send of the file,
+  and simply replies with the error messages and codes from the failed upload-info sign attempt."
+  [upload-info ch]
+  (if (contains? upload-info :error-code)
+    (do
+      (put! ch (select-keys upload-info [:identifier :error-code :error-message :http-error-code]))
+      (close! ch))
+    (let [xhr        (XhrIo.)
+          identifier (:identifier upload-info)
+          sig-fields [:key :Content-Type :success_action_status :policy :AWSAccessKeyId :signature :acl]
+          signature  (select-keys (:signature upload-info) sig-fields)
+          form-data  (formdata-from-map (merge signature {:file (:f upload-info)}))]
+      (events/listen xhr goog.net.EventType.SUCCESS
+        (fn [_]
+          (put! ch {:file       (:f upload-info)
+                    :response   (if-let [response-xml (.getResponseXml xhr)]
+                                    (xml->map response-xml)
+                                    "")
+                    :xhr xhr
+                    :identifier identifier})
+          (close! ch)))
+      (events/listen xhr goog.net.EventType.ERROR
+        (fn [_]
+          (let [error-code      (.getLastErrorCode xhr)
+                error-message   (.getDebugMessage goog.net.ErrorCode error-code)
+                http-error-code (.getStatus xhr)]
+            (put! ch {:identifier      identifier
+                      :error-code      error-code
+                      :error-message   (str "While trying to upload file: "
+                                             error-message)
+                      :response   (if-let [response-xml (.getResponseXml xhr)]
+                                    (xml->map response-xml)
+                                    "")
```

Comment:
I’ve fixed this.

http://www.getcontented.com.au/ - Amazing website creation service. Get a professional, distinguished yet highly affordable website today.
http://www.getcontented.com.au/flute_websites.html - website design & hosting service for flutists








> On 3 Sep 2015, at 8:41 pm, Martin Klepsch <notifications@github.com> wrote:
> 
> In src/cljs/s3_beam/client.cljs <https://github.com/martinklepsch/s3-beam/pull/21#discussion_r38632579>:
> 
> > +                                    "")
> > +                    :xhr xhr
> > +                    :identifier identifier})
> > +          (close! ch)))
> > +      (events/listen xhr goog.net.EventType.ERROR
> > +        (fn [_]
> > +          (let [error-code      (.getLastErrorCode xhr)
> > +                error-message   (.getDebugMessage goog.net.ErrorCode error-code)
> > +                http-error-code (.getStatus xhr)]
> > +            (put! ch {:identifier      identifier
> > +                      :error-code      error-code
> > +                      :error-message   (str "While trying to upload file: "
> > +                                             error-message)
> > +                      :response   (if-let [response-xml (.getResponseXml xhr)]
> > +                                    (xml->map response-xml)
> > +                                    "")
> I think this would be the most idiomatic way:
> 
> (when-let [response-xml (.getResponseXml xhr)]
>   (xml->map response-xml))
> —
> Reply to this email directly or view it on GitHub <https://github.com/martinklepsch/s3-beam/pull/21/files#r38632579>.
>